### PR TITLE
Enchantment/relic cleanup 1

### LINKED ISSUE #186: Intermittent_activation relic effects break on save/load
# Describe the bug

Relics with `intermittent_activation` enchantments take effect normally up until the game is saved, and then loaded. Afterward, the effects will cease to activate.

This appears to be a separate issue from the "intermittent activation relic_data specified directly inside the item breaks saves" bug in that it takes affect even if `enchantment::serialize` is fixed, and can be tested with explicitly-defined enchantments (which avoid the save-breaking bug).

# Steps To Reproduce

1. Use version of branch from https://github.com/cataclysmbnteam/Cataclysm-BN/pull/185, which fixes the save breakages and updates the Cube of Shame to be usable for testing this function.
2. Start game and spawn in a Cube of Shame.
3. Wait 5-10 minutes for healing messages to appear, even at full HP.
4. Save and load game.
5. Resume waiting. No more messages from the spell triggering.

# Expected behavior

Save/load to have no impact on function of intermittent_activation.

# Screenshots

Before saving:
image

After saving and loading (in a test build for the PR fixing save breakage):
image

# Versions and configuration

# Additional context

```
- OS: Windows
    - OS Version: 10.0 1803
- Game Version: 7da6104589-dirty [64-bit]
- Graphics Version: Tiles
- Game Language: System language []
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food]
]
```

Related DDA issue: https://github.com/CleverRaven/Cataclysm-DDA/issues/41261

As mentioned, it seems that fixing the save breakage is easy enough, but it simply exposed an underlying issue, which I was able to eventually confirm affects sources of `intermittent_activation` that would not normally be subject to the save/load bug.

### PR DESCRIPTION
**Pull request by olanti-p**
#### Purpose of change
1. Remove magic-related headers from commonly included headers (recompiling is a pain otherwise)
2. Don't stack items with different relic_data
3. Fix #186 and another similar bug, with effects-from-enchantments
4. When saving game map, don't write `"relic_data": null` for every single item
5. JSONize The Architect's Cube
6. Update and expand enchantments documentation

#### Describe the solution
See commits. Some of the changes were cherry-picked from DDA repo, but due to merge conflicts and/or fixes I had to apply the final code is a bit of a mess, authorship-wise. Pretty sure I've credited everyone properly though.

#### Describe alternatives you've considered
Docs could go on the wiki once everything's up to date?

#### Testing
Manual for now. Checked that Cube of Shame survives save/load with it's `relic_data` intact, tried out the Architect's Cube.
Automated tests will have to wait until enchantment modifiers are implemented.

### CODE REVIEW

File: src/magic.cpp [hunk 1]
```diff
@@ -2050,6 +2050,17 @@ spell fake_spell::get_spell( int min_level_override ) const
     return sp;
 }
 
+bool fake_spell::operator==( const fake_spell &rhs )const
```
Comment (Coolthulhu) [inline src/magic.cpp:2053]: Astyle is fine with that un-spaced `const`?

**Comment by Coolthulhu** [comment]
> When saving game map, don't write "relic_data": null for every single item

That's like 50% of useless item save data gone. Now it's just `"last_rot_check": 0` on non-food items and many of them will only have the type.

File: src/magic.cpp (hunk 1 above)
Comment (olanti-p) [reply src/magic.cpp:2053]: Apparently yes...
```suggestion
bool fake_spell::operator==( const fake_spell &rhs ) const
```

**Comment by olanti-p** [comment]
That reminds me, are there any plans on moving to c++17 or even c++20?

**Comment by Coolthulhu** [comment]
Not at the moment, but if you need something from those for important changes (or changes that would otherwise be huge), it may be worth considering.

**Comment by Coolthulhu** [comment]
For some reason, this breaks vision tests on my local build.
It's weird, because none of the Travis builds broke.
I'll try again tomorrow.

**Comment by Coolthulhu** [comment]
Alright, it wasn't this PR, it was my fault. This PR looks fine.

**Pull request merged**